Translate "Forms" (#174)
reactjs/ru.reactjs.org: content/docs/forms.md

  - "docs/forms-zh-CN.html"
---

> В React HTML элементы формы ведут себя несколько отлично от остальных DOM-элементов, так как у элементов формы изначально есть внутреннее состояние. К примеру, в HTML форму ниже можно ввести имя:

Reviewer: Обратите внимание на написание сложных существительных, таких как HTML-элементы, их нужно [писать через дефис](http://new.gramota.ru/biblio/readingroom/rules/133-def-19-19).

```suggestion
В React HTML-элементы формы ведут себя несколько отлично от остальных DOM-элементов, так как у элементов формы изначально есть внутреннее состояние. К примеру, в HTML-форму ниже можно ввести имя:
```

```html
<form>
  <label>
    Имя:
    <input type="text" name="name" />
  </label>
  <input type="submit" value="Отправить" />
</form>
```

> При отправке пользователем, поведением по умолчанию для такой HTML формы будет переход на новую страницу. Точно так же она будет вести себя и в React. Однако, чаще всего удобно обрабатывать отправку формы через JavaScript функцию, которая имеет доступ к введённым в форму данным. Для этого можно воспользоваться стандартным способом под названием "управляемые компоненты".

Reviewer: ```suggestion
При отправке пользователем, поведением по умолчанию для такой HTML-формы будет переход на новую страницу. Точно так же она будет вести себя и в React. Однако, чаще всего удобно обрабатывать отправку формы через JavaScript-функцию, которая имеет доступ к введённым в форму данным. Для этого можно воспользоваться стандартным способом под названием «управляемые компоненты».
```

## Управляемые компоненты {#controlled-components}

Такие HTML элементы формы, как `<input>`, `<textarea>`, and `<select>` обычно сами управляют своим состоянием и обновляют его когда пользователь вводит данные. В React, мутабельное состояние обычно содержится в свойстве компонентов `state` и обновляется только через вызов[`setState()`](/docs/react-component.html#setstate)

> Мы можем скомбинировать оба подхода и сделать состояние React компонента "единственным источником правды". Тогда React компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться "управляемый компонент".

Author: Не уверен насчет необходимости дефиса между React и компонент. Возможно, я упускаю что то?

> Мы можем скомбинировать оба подхода и сделать состояние React компонента "единственным источником правды". Тогда React компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться "управляемый компонент".

Author: Хотя, если DOM-элемент, то и React-component

> Мы можем скомбинировать оба подхода и сделать состояние React компонента "единственным источником правды". Тогда React компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться "управляемый компонент".

Reviewer 2: В [TRANSLATION.md](https://github.com/reactjs/ru.reactjs.org/blob/master/TRANSLATION.md#%D1%81%D0%BE%D0%B3%D0%BB%D0%B0%D1%88%D0%B5%D0%BD%D0%B8%D0%B5-%D0%BF%D0%BE-%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4%D1%83-%D0%B3%D0%BB%D0%BE%D1%81%D1%81%D0%B0%D1%80%D0%B8%D0%B9)

> React element |  React-элемент, элемент React

По аналогии и все остальные

> Мы можем скомбинировать оба подхода и сделать состояние React-компонента "единственным источником правды". Тогда React-компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться "управляемый компонент".

Reviewer 2: An input form element - не в смысле `input`, а в смысле _элемент для ввода данных_ или просто _Элемент формы_.

```suggestion
Мы можем объединить оба подхода, если сделаем состояние React-компонента «единственным источником истины». Тогда React-компонент, который рендерит форму, будет ещё и реагировать на ввод данных в форму. Элемент формы, значением которого теперь управляет React, будет называться «управляемым компонентом»
```

Author: Подожду, что ответят в переводе uncontrolled components

Reviewer 2: Можно создать ишью про этот термин

Author: До завтра жду и создам

> Мы можем скомбинировать оба подхода и сделать состояние React-компонента "единственным источником правды". Тогда React-компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться "управляемый компонент".

Reviewer: ```suggestion
Мы можем скомбинировать оба подхода и сделать состояние React-компонента "единственным источником правды". Тогда React-компонент будет рендерить форму и управлять её поведением в ответ на пользовательский ввод. Значением элемента формы input в этом случае будет управлять React, а сам элемент будет называться «управляемый компонент».
```

> Допустим, требуется, чтобы предыдущий пример при отправке выводил в лог имя. Тогда можно написать форму в виде управляемого компонента:

Reviewer 2: ```suggestion
Допустим, мы хотим, чтобы предыдущий пример выводил в лог имя, когда мы отправляем форму. Тогда можно написать форму в виде управляемого компонента:
```

> Допустим, требуется, чтобы предыдущий пример при отправке выводил в лог имя. Тогда можно написать форму в виде управляемого компонента:

Reviewer: ```suggestion
Допустим, требуется, чтобы предыдущий пример при отправке выводил в консоль имя. Тогда можно написать форму в виде управляемого компонента:
```

  }

  handleSubmit(event) {
    alert('Отправленное имя: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Имя:
          <input type="text" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Отправить" />
      </form>
    );
  }
}
```

[**Посмотреть на CodePen**](https://codepen.io/gaearon/pen/VmmPgp?editors=0010)

>  На элементе формы input установлен атрибут `value`, а его отображаемое значение всегда будет `this.state.value`, что делает состояние React-компонента "источником правды". А, так как каждое нажатие клавиши вызывает `handleChange`, который обновляет состояние React-компонента, отображаемое значение будет обновляться по мере того как пользователь печатает.

Reviewer 2: ```suggestion
Мы установили атрибут `value` для поля ввода, и теперь в нём всегда будет отображаться `this.state.value`. Состояние React-компонента стало «источником истины». А, так как каждое нажатие клавиши вызывает `handleChange`, который обновляет состояние React-компонента, значение в поле будет обновляться по мере того как пользователь печатает.
```

>  На элементе формы input установлен атрибут `value`, а его отображаемое значение всегда будет `this.state.value`, что делает состояние React-компонента "источником правды". А, так как каждое нажатие клавиши вызывает `handleChange`, который обновляет состояние React-компонента, отображаемое значение будет обновляться по мере того как пользователь печатает.

Reviewer: ```suggestion
На элементе формы input установлен атрибут `value`, а его отображаемое значение всегда будет `this.state.value`, что делает состояние React-компонента "источником правды". А, так как каждое нажатие клавиши вызывает `handleChange`, который обновляет состояние React-компонента, отображаемое значение будет обновляться по мере того, как пользователь печатает.
```

> В управляемом компоненте с каждой мутацией состояния связана функция-обработчик. Это делает валидацию или изменение пользовательского ввода простой задачей. Например, если требуется, чтобы имя обязательно состояло из заглавных букв, то можно написать такой `handleChange`:

Reviewer 2: ```suggestion
В управляемом компоненте с каждой мутацией состояния связана функция-обработчик. Благодаря этому валидация или изменение введённого значения становится простой задачей. Например, если мы хотим, чтобы имя обязательно было набрано заглавными буквами, можно написать такой `handleChange`:
```

```javascript{2}
handleChange(event) {
  this.setState({value: event.target.value.toUpperCase()});
}
```

> ## Тэг textarea {#the-textarea-tag}

Reviewer: ```suggestion
## Тег textarea {#the-textarea-tag}
```

Author: Для целостности обновил, но я лично считаю что правильнее было бы писать тэг, чтобы поощрять корректное произношение, соотносящееся с английским вариантом. Люди, говорящие "тег", встречаются не так редко, как хотелось бы.

HTML-элемент `<textarea>` в качестве текста отображает дочерний элемент:

```html
<textarea>
  Доброго здоровья, тут просто немного текста внутри тега textarea
</textarea>
```

> В React `<textarea>` использует атрибут `value`. Таким образом, форму с `<textarea>` можно написать почти тем же способом что и форму с однострочным `<input>`: 

Reviewer: ```suggestion
В React `<textarea>` использует атрибут `value`. Таким образом, форму с `<textarea>` можно написать почти тем же способом, что и форму с однострочным `<input>`: 
```

  }

  handleSubmit(event) {
    alert('Сочинение отправлено: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Сочинение:
          <textarea value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Отправить" />
      </form>
    );
  }
}
```

> Обратите внимание, что `this.state.value` инициализировано в конструкторе, так что в текстовой области изначально есть текст.

Reviewer 2: ```suggestion
Обратите внимание, что мы инициализировали `this.state.value` в конструкторе, так что в текстовой области изначально есть текст.
```

> ## Тэг select {#the-select-tag}

Reviewer: ```suggestion
## Тег select {#the-select-tag}
```

В HTML `<select>` создаёт выпадающий список. HTML-код в этом примере создаёт выпадающий список вкусов:

```html
<select>
  <option value="grapefruit">Грейпфрут</option>
  <option value="lime">Лайм</option>
  <option selected value="coconut">Кокос</option>
  <option value="mango">Манго</option>
</select>
```

> Опция "Кокос" выбрана по умолчанию из-за установленного атрибута `selected`. React вместо этого атрибута использует `value` в корневом теге `select`. В управляемом компоненте так удобнее, потому что обновлять значение нужно только в одном месте (`state`). Пример:

Reviewer: ```suggestion
Вариант «Кокос» выбран по умолчанию из-за установленного атрибута `selected`. React вместо этого атрибута использует `value` в корневом теге `select`. В управляемом компоненте так удобнее, потому что обновлять значение нужно только в одном месте (`state`). Пример:
```

Author: Думаю, "Пункт списка" будет лучше.

> Опция "Кокос" выбрана по умолчанию из-за установленного атрибута `selected`. React вместо этого атрибута использует `value` в корневом теге `select`. В управляемом компоненте так удобнее, потому что обновлять значение нужно только в одном месте (`state`). Пример:

Reviewer 2: ```suggestion
Опция "Кокос" выбрана по умолчанию из-за установленного атрибута `selected`. React вместо этого атрибута использует `value` в корневом теге `select`. В управляемом компоненте так удобнее, потому что обновлять значение нужно только в одном месте. Пример:
```

Author: Считаю, что лучше это указание оставить. Его наличие никакого негативного эффекта не несет, но в случае вопроса дает разъяснение.

  }

  handleSubmit(event) {
    alert('Ваш любимый вкус: ' + this.state.value);
    event.preventDefault();
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Выберите ваш любимый вкус:
          <select value={this.state.value} onChange={this.handleChange}>
            <option value="grapefruit">Грейпфрут</option>
            <option value="lime">Лайм</option>
            <option value="coconut">Кокос</option>
            <option value="mango">Манго</option>
          </select>
        </label>
        <input type="submit" value="Отправить" />
      </form>
    );
  }
}
```

[**Посмотреть на CodePen**](https://codepen.io/gaearon/pen/JbbEzX?editors=0010)

Подводя итог, `<input type="text">`, `<textarea>`, и `<select>` работают очень похоже. Все они принимают атрибут `value`, который можно использовать, чтобы реализовать управляемый компонент.

> Примечание
>

> > Вы можете передать массив внутрь атрибута `value`, что позволит выбрать несколько опций в теге `select`:

Reviewer 2: ```suggestion
> В атрибут `value` можно передать массив, что позволит выбрать несколько опций в теге `select`:
```

>
>```js
><select multiple={true} value={['Б', 'В']}>
>```

## Загрузка файла {#the-file-input-tag}

> ## The file input Tag {#the-file-input-tag}

Reviewer: ```suggestion

```

> В HTML `<input type="file">` позволяет пользователю выбрать один или несколько файлов для загрузки с устройства на сервер или обработки с помощью [File API](https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications) JavaScript. 

Reviewer: ```suggestion
В HTML `<input type="file">` позволяет пользователю выбрать один или несколько файлов для загрузки с устройства на сервер или управлять им через JavaScript с помощью [File API](https://developer.mozilla.org/ru/docs/Web/API/File/Using_files_from_web_applications
).
```

```html
<input type="file" />
```

Так как значение такого элемента доступно только для чтения, это **неуправляемый** React-компонент. Мы расскажем про этот и другие неуправляемые компоненты [далее в документации](/docs/uncontrolled-components.html#the-file-input-tag).

## Обработка нескольких input {#handling-multiple-inputs}

> Когда вы имеете дело с несколькими управляемыми элементами `input`, вы можете добавить каждому из них атрибут `name`, что позволит функции-обработчику решать, что делать, основываясь на значении `event.target.name`.

Reviewer 2: ```suggestion
Если вам нужны несколько управляемых элементов `input`, вы можете назначить каждому из них атрибут `name`, что позволит функции-обработчику решать, что делать, основываясь на значении `event.target.name`.
```

}
```

[**Посмотреть на CodePen**](https://codepen.io/gaearon/pen/wgedvV?editors=0010)

> Обратите внимание, мы используем синтаксиса ES6 [вычисляемые имена](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names), чтобы обновить значение в `state` через ключ, который соответствует атрибуту name элемента input:

Reviewer: ```suggestion
Обратите внимание, мы используем ES6-возможность [вычисляемые имена свойств](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Operators/Object_initializer#%D0%92%D1%8B%D1%87%D0%B8%D1%81%D0%BB%D1%8F%D0%B5%D0%BC%D1%8B%D0%B5_%D0%B8%D0%BC%D0%B5%D0%BD%D0%B0_%D1%81%D0%B2%D0%BE%D0%B9%D1%81%D1%82%D0%B2), чтобы обновить значение в `state` через ключ, который соответствует атрибуту name элемента input:
```

> Обратите внимание, мы используем синтаксиса ES6 [вычисляемые имена](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names), чтобы обновить значение в `state` через ключ, который соответствует атрибуту name элемента input:

Reviewer 2: ```suggestion
Обратите внимание, что мы используем [вычисляемые имена](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names) из синтаксиса ES6, чтобы обновить значение в `state` через ключ, соответствующий названию поля для ввода:
```

```js{2}
this.setState({
  [name]: value
});
```

Идентичный ES5 код:

```js{2}
var partialState = {};
partialState[name] = value;
this.setState(partialState);
```

> Кроме того, `setState()` автоматически производит [слияние части состояния с текущим состоянием](/docs/state-and-lifecycle.html#state-updates-are-merged), то есть можно вызывать этот метод только для изменяемых в данный момент частей `state`.

Reviewer 2: ```suggestion
Кроме того, `setState()` автоматически производит [слияние части состояния с текущим состоянием](/docs/state-and-lifecycle.html#state-updates-are-merged), то есть в него можно передавать только данные, которые изменились.
```

Author: По поему, это меняет смысл на "передача данных, которые не изменились, запрещена"

Reviewer 2: Да, можно так истолковать. Но это проблема сочетания _можно только_, как будто остальное нельзя, мне кажется, такое толкование возможно и в вашем варианте.
Может добавить в конце пояснение?

> то есть в него можно передавать только данные, которые изменились, а остальные он возьмёт из текущего `state`.

Или можно сделать предложение личным, как в оригинале:

> то есть нам нужно передать в него только ту часть `state` , которую хотим изменить.

Тогда, _нужно_ может и лучше будет звучать. Выходит, что остальное _не нужно_.

Author: Второе 👍

## Значение null управляемого компонента {#controlled-input-null-value}

> Передача пропа value [управляемому компоненту](/docs/forms.html#controlled-components) позволит вам не дать пользователю изменить значение input без вашего желания. Если значение `value` передано, но input остается редактируемым, возможно, значение `value` случайно было установлено в `undefined` или `null`.

Reviewer 2: ```suggestion
Если установить [управляемому компоненту](/docs/forms.html#controlled-components) проп `value`, то пользователь не сможет изменить его значение без вашего желания. Если вы установили `value`, а поле ввода по-прежнему можно редактировать, то, возможно, вы случайно задали `value`, равный `undefined` или `null`.
```

> Код ниже это демонстрирует. (после небольшой задержки, сперва заблокированный для ввода input становится редактируемым.)

Reviewer 2: ```suggestion
Код ниже это демонстрирует. (Изначально заблокированный `input` становится редактируемым после небольшой задержки.)
```

```javascript
ReactDOM.render(<input value="Привет" />, mountNode);

setTimeout(function() {
  ReactDOM.render(<input value={null} />, mountNode);
}, 1000);

```

## Альтернативы управляемым компонентам {#alternatives-to-controlled-components}

> Использование управляемых компонентов иногда может быть утомительным. Приходится писать обработчик события для каждого варианта изменения ваших данных и проводить всё состояние input через React компонент. Это может особенно раздражать если вы переносите существующую кодовую базу в React, или когда работаете над интеграцией React приложения с не—React библиотекой. В такой ситуации могут пригодиться [uncontrolled components](/docs/uncontrolled-components.html), альтернативная техника реализации ввода данных в форму.

Reviewer 2: ```suggestion
Использование управляемых компонентов иногда может быть утомительным. Приходится писать обработчик события для каждого варианта изменения ваших данных и проводить всё состояние input через React компонент. Это может особенно раздражать если вы переносите существующую кодовую базу в React, или когда работаете над интеграцией React приложения с другой библиотекой. В такой ситуации могут пригодиться [uncontrolled components](/docs/uncontrolled-components.html), альтернативная техника реализации ввода данных в форму.
```

> Использование управляемых компонентов иногда может быть утомительным. Приходится писать обработчик события для каждого варианта изменения ваших данных и проводить всё состояние input через React-компонент. Это может особенно раздражать, если вы переносите существующую кодовую базу в React, или когда работаете над интеграцией React приложения с другой библиотекой. В такой ситуации могут пригодиться [неуправляемые компоненты](/docs/uncontrolled-components.html), альтернативная техника реализации ввода данных в форму.

Reviewer 2: ```suggestion
Использование управляемых компонентов иногда может быть утомительным. Приходится писать обработчик события для каждого варианта изменения ваших данных и проводить всё состояние формы через компонент React. Это может особенно раздражать, если вы переносите существующую кодовую базу в React, или когда работаете над интеграцией React приложения с другой библиотекой. В такой ситуации могут пригодиться [неуправляемые компоненты](/docs/uncontrolled-components.html), альтернативная техника реализации ввода данных в форму.
```

## Полноценные решения {#fully-fledged-solutions}

> При поиске полноценного решения, которое может валидировать ввод, запомнить посещённые поля формы и обработать ее отправку, [Formik](https://jaredpalmer.com/formik) будет одной из популярных опций. Как бы то ни было, эта библиотека построена на принципах управляемых компонентов и управления состоянием, так что не пренебрегайте их изучением.

Reviewer 2: ```suggestion
Ели вы ищите полноценное решение, которое может валидировать ввод, запомнить посещённые поля формы и обработать ее отправку, присмотритесь к [Formik](https://jaredpalmer.com/formik). Как бы то ни было, эта библиотека построена на принципах управляемых компонентов и управления состоянием, так что не пренебрегайте их изучением.
```
